### PR TITLE
modify test path - module gatk4   +  added new module gatk4/haplotypecaller

### PR DESCRIPTION
**Pull request by suzannejin**
Modified test paths:
- [x] gatk4/createsequencedictionary (@suzannejin)
- [x] gatk4/bedtointervallist (@suzannejin)
- [x] gatk4/samtofastq (@suzannejin)
- [x] gatk4/splitncigarreads (@suzannejin)
- [x] gatk4/fastqtosam (@suzannejin)
- [x] gatk4/revertsam (@suzannejin)
- [x] gatk4/mergebamalignment (@suzannejin)
- [x] gatk4/applybqsr (@suzannejin)
- [x] gatk4/baserecalibrator (@suzannejin)
- [x] gatk4/mergevcfs (@suzannejin)

And added new module: gatk4/haplotypecaller

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [ ] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Review by rivera10 — APPROVED** [review]
LGTM

File: software/gatk4/haplotypecaller/main.nf [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/gatk4:4.2.0.0--0"
+    } else {
+        container "quay.io/biocontainers/gatk4:4.2.0.0--0"
+    }
+
+    input:
+    tuple val(meta), path(bam), path(bamidx)
+    path fasta
+    path fastaidx
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:24]:
```suggestion
    path fai
```

File: software/gatk4/haplotypecaller/main.nf [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    path dict
+
+    output:
+    tuple val(meta), path("*.vcf.gz"), emit: vcf
+    path("*.vcf.gz.tbi")             , emit: vcfidx
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:29]:
```suggestion
    path("*.vcf.gz.tbi")             , emit: tbi
```

File: software/gatk4/haplotypecaller/main.nf [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    path "*.version.txt"             , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:35]:
```suggestion
```

File: software/gatk4/haplotypecaller/main.nf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    gatk --java-options "-Xmx4g" HaplotypeCaller \\
+        -R $fasta \\
+        -I $bam \\
+        -O ${prefix}.vcf.gz \\
+        $options.args
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:41]:
```suggestion
    gatk \\
        --java-options "-Xmx4g" \\
        HaplotypeCaller \\
        -R $fasta \\
        -I $bam \\
        -O ${prefix}.vcf.gz \\
        $options.args
```

Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:41]:
```suggestion
    def software  = getSoftwareName(task.process)
    def prefix    = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
    def avail_mem = 3
    if (!task.memory) {
        log.info '[GATK HaplotypeCaller] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = task.memory.giga
    }
    """
    gatk \\
        --java-options "-Xmx${avail_mem}g" \\
        HaplotypeCaller \\
        -R $fasta \\
        -I $bam \\
        -O ${prefix}.vcf.gz \\
        $options.args
        
```

Comment (drpatelh) [reply software/gatk4/haplotypecaller/main.nf:41]: We shouldn't hard-code the memory requirements for the process. It should ideally be set by `task.memory`

File: software/gatk4/haplotypecaller/main.nf [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' > ${software}.version.txt
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:43]: `samtools` version?

File: software/gatk4/haplotypecaller/meta.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      pattern: "*.bam"
+  - bamidx:
+      type: file
+      description: Index of BAM file
+      pattern: "*.bam.bai"
+  - fasta:
+      type: file
+      description: The reference fasta file
+      pattern: "*.fasta"
+  - fastaidx:
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/meta.yml:35]:
```suggestion
  - fai
```

File: software/gatk4/haplotypecaller/meta.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - vcf:
+      type: file
+      description: Compressed VCF file
+      pattern: "*.vcf.gz"
+  - vcfidx:
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/meta.yml:57]:
```suggestion
  - tbi
```

File: tests/software/gatk4/haplotypecaller/main.nf [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { GATK4_HAPLOTYPECALLER } from '../../../../software/gatk4/haplotypecaller/main.nf' addParams( options: [:] )
+
+workflow test_gatk4_haplotypecaller {
+    
+    def input = []
```
Comment (drpatelh) [inline tests/software/gatk4/haplotypecaller/main.nf:9]:
```suggestion
```

Comment (drpatelh) [reply tests/software/gatk4/haplotypecaller/main.nf:9]: Don't really need this. Will get it removed from the modules template too at some point.

File: tests/software/gatk4/haplotypecaller/test.yml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+- name: gatk4 haplotypecaller test_gatk4_haplotypecaller
+  command: nextflow run tests/software/gatk4/haplotypecaller -entry test_gatk4_haplotypecaller -c tests/config/nextflow.config
+  tags:
+    - gatk4_haplotypecaller
+    - gatk4
+  files:
+    - path: output/gatk4/test.vcf.gz
+      should_exist: true
+      contains:
+        - 'MT192765.1'
```
Comment (drpatelh) [inline tests/software/gatk4/haplotypecaller/test.yml:10]: Nice!

File: tests/software/gatk4/haplotypecaller/test.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+        - '54.60'
+        - '37.32'
+    - path: output/gatk4/test.vcf.gz.tbi
+      should_exist: true
```
Comment (drpatelh) [inline tests/software/gatk4/haplotypecaller/test.yml:14]:
```suggestion
```

Comment (drpatelh) [reply tests/software/gatk4/haplotypecaller/test.yml:14]: By default, I think `pytest` will check that the file exists if you provide `path`

**Review by drpatelh — COMMENTED** [review]
Thank you @suzannejin ! Some minor changes. I am a bit concerned as to why we needed to change all of those md5sums given that I haven't changed the test data files themselves? Do the output files have timestamps or other things that prevent us from using an md5sum e.g the header of the BAM files? 

One way to test would be to try and re-run the tests locally for those modules and see if they fail again because something may have changed in the output file.

**Comment by suzannejin** [comment]
> Thank you @suzannejin ! Some minor changes. I am a bit concerned as to why we needed to change all of those md5sums given that I haven't changed the test data files themselves? Do the output files have timestamps or other things that prevent us from using the same md5sum e.g the header of the BAM files?
> 
> One way to test would be to try and re-run the tests locally for those modules and see if they fail again because something may have changed in the output file.

Hii!! Sorry for the delay in replying. The test.vcf.gz file from the haplotypecaller has a date stamp inside, so that's why md5sum changes. But I could not find the reason why my md5sum changed for eg. applybqsr compared to the original md5sum, since when I rerun it, the md5sum remained the same one for me.

File: software/gatk4/haplotypecaller/meta.yml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      pattern: "*.bam"
+  - bamidx:
+      type: file
+      description: Index of BAM file
+      pattern: "*.bam.bai"
+  - fasta:
+      type: file
+      description: The reference fasta file
+      pattern: "*.fasta"
+  - fai
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/meta.yml:35]:
```suggestion
  - fai:
```

File: software/gatk4/haplotypecaller/main.nf (hunk 5 above)
Comment (drpatelh) [reply software/gatk4/haplotypecaller/main.nf:51]: This still needs to be updated?

File: software/gatk4/haplotypecaller/main.nf [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    """
+    gatk \\
+        --java-options "-Xmx${avail_mem}g" \\
+        HaplotypeCaller \\
+        -R $fasta \\
+        -I $bam \\
+        -O ${prefix}.vcf.gz \\
+        $options.args
+
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:49]:
```suggestion
```

File: software/gatk4/haplotypecaller/meta.yml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - vcf:
+      type: file
+      description: Compressed VCF file
+      pattern: "*.vcf.gz"
+  - tbi
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/meta.yml:57]:
```suggestion
  - tbi:
```

File: software/gatk4/haplotypecaller/main.nf [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    conda (params.enable_conda ? "bioconda::gatk4=4.2.0.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/gatk4:4.2.0.0--0"
+    } else {
+        container "quay.io/biocontainers/gatk4:4.2.0.0--0"
+    }
+
+    input:
+    tuple val(meta), path(bam), path(bamidx)
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:22]:
```suggestion
    tuple val(meta), path(bam), path(bai)
```

File: software/gatk4/haplotypecaller/main.nf [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    path fasta
+    path fai
+    path dict
+
+    output:
+    tuple val(meta), path("*.vcf.gz"), emit: vcf
+    path("*.vcf.gz.tbi")             , emit: tbi
+    path "*.version.txt"             , emit: version
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:30]:
```suggestion
    tuple val(meta), path("*.vcf.gz"), emit: vcf
    tuple val(meta), path("*.tbi")   , emit: tbi
    path "*.version.txt"             , emit: version
```

File: software/gatk4/haplotypecaller/main.nf [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    gatk \\
+        --java-options "-Xmx${avail_mem}g" \\
+        HaplotypeCaller \\
+        -R $fasta \\
+        -I $bam \\
+        -O ${prefix}.vcf.gz \\
+        $options.args
+        
+
+    echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' > ${software}.version.txt
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/main.nf:51]:
```suggestion
    gatk --version | grep Picard | sed "s/Picard Version: //g" > ${software}.version.txt
```

File: software/gatk4/haplotypecaller/meta.yml [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - bam:
+      type: file
+      description: BAM file
+      pattern: "*.bam"
+  - bamidx:
```
Comment (drpatelh) [inline software/gatk4/haplotypecaller/meta.yml:27]:
```suggestion
  - bai:
```

**Comment by drpatelh** [comment]
Thanks for checking the md5sums @suzannejin ! I pushed a commit changing some other stuff and fixing the linting. Feel to merge whenever you are happy 🙂

**Review by drpatelh — APPROVED** [review]

**Comment by suzannejin** [comment]
> Thanks for checking the md5sums @suzannejin ! I pushed a commit changing some other stuff and fixing the linting. Feel to merge whenever you are happy 🙂

Thanks to you @drpatelh for reviewing everything!!!

**Pull request merged**